File: Data Structures & Algorithms/trapping-rain-water/submission-2.js
class Solution {
    /**
     * @param {number[]} height
     * @return {number}
     */
    trap(height) {
        const size = height.length;
        let water = 0;
        // for(let i=0;i<size;i++){
        //     let leftMax=0;
        //     let rightMax=0;

        //     for(let l=0;l<=i;l++){
        //         leftMax = Math.max(height[l],leftMax);
        //     }

        //     for(let r=i;r<size;r++){
        //         rightMax = Math.max(height[r],rightMax);
        //     }

        //     water += Math.min(leftMax,rightMax) - height[i];
        // }

        let left = 0;
        let right = size-1;
        let leftMax=0,rightMax=0;
        while(left<right){
            if(height[left] <height[right]){
                if(height[left]>=leftMax){
                    leftMax = height[left];
                }else{
                    water+= leftMax - height[left];
                }
                left++;
            }else{
                if(height[right]>=rightMax){
                    rightMax = height[right];
                }else{
                    water+= rightMax - height[right];
                }
                right--;
            }
        }
        return water;

    }
}

File: Data Structures & Algorithms/max-water-container/submission-2.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {

        int left  = 0;
        int right = heights.size()-1;
        int maxArea=0;
        while(left<right){
            int area = min(heights[left],heights[right]) * (right-left);
            maxArea = max(area,maxArea);
            if(heights[left]>=heights[right]){
                right--;
            }else if(heights[left]<heights[right]){
                left++;
            }
        }
        
        return maxArea;
    }
};
